{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Loading required package: plyr\n",
      "\n",
      "Loading required package: ggplot2\n",
      "\n",
      "Loading required package: Rtsne\n",
      "\n",
      "Loading required package: grid\n",
      "\n",
      "Loading required package: reshape\n",
      "\n",
      "\n",
      "Attaching package: ‘reshape’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:plyr’:\n",
      "\n",
      "    rename, round_any\n",
      "\n",
      "\n",
      "Loading required package: RColorBrewer\n",
      "\n",
      "Loading required package: Biobase\n",
      "\n",
      "Loading required package: BiocGenerics\n",
      "\n",
      "Loading required package: parallel\n",
      "\n",
      "\n",
      "Attaching package: ‘BiocGenerics’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:parallel’:\n",
      "\n",
      "    clusterApply, clusterApplyLB, clusterCall, clusterEvalQ,\n",
      "    clusterExport, clusterMap, parApply, parCapply, parLapply,\n",
      "    parLapplyLB, parRapply, parSapply, parSapplyLB\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    IQR, mad, sd, var, xtabs\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    anyDuplicated, append, as.data.frame, basename, cbind, colnames,\n",
      "    dirname, do.call, duplicated, eval, evalq, Filter, Find, get, grep,\n",
      "    grepl, intersect, is.unsorted, lapply, Map, mapply, match, mget,\n",
      "    order, paste, pmax, pmax.int, pmin, pmin.int, Position, rank,\n",
      "    rbind, Reduce, rownames, sapply, setdiff, sort, table, tapply,\n",
      "    union, unique, unsplit, which, which.max, which.min\n",
      "\n",
      "\n",
      "Welcome to Bioconductor\n",
      "\n",
      "    Vignettes contain introductory material; view with\n",
      "    'browseVignettes()'. To cite Bioconductor, see\n",
      "    'citation(\"Biobase\")', and for packages 'citation(\"pkgname\")'.\n",
      "\n",
      "\n",
      "Loading required package: ape\n",
      "\n",
      "Loading required package: MASS\n",
      "\n",
      "Loading required package: stringi\n",
      "\n",
      "Loading required package: roxygen2\n",
      "\n",
      "Loading required package: phytools\n",
      "\n",
      "Loading required package: maps\n",
      "\n",
      "\n",
      "Attaching package: ‘maps’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:plyr’:\n",
      "\n",
      "    ozone\n",
      "\n",
      "\n",
      "Loading required package: repr\n",
      "\n",
      "Loading required package: IRanges\n",
      "\n",
      "Loading required package: S4Vectors\n",
      "\n",
      "Loading required package: stats4\n",
      "\n",
      "\n",
      "Attaching package: ‘S4Vectors’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:reshape’:\n",
      "\n",
      "    expand, rename\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:plyr’:\n",
      "\n",
      "    rename\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:base’:\n",
      "\n",
      "    expand.grid\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘IRanges’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:plyr’:\n",
      "\n",
      "    desc\n",
      "\n",
      "\n",
      "Loading required package: SummarizedExperiment\n",
      "\n",
      "Loading required package: GenomicRanges\n",
      "\n",
      "Loading required package: GenomeInfoDb\n",
      "\n",
      "Loading required package: DelayedArray\n",
      "\n",
      "Loading required package: matrixStats\n",
      "\n",
      "\n",
      "Attaching package: ‘matrixStats’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:Biobase’:\n",
      "\n",
      "    anyMissing, rowMedians\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:plyr’:\n",
      "\n",
      "    count\n",
      "\n",
      "\n",
      "Loading required package: BiocParallel\n",
      "\n",
      "\n",
      "Attaching package: ‘DelayedArray’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:matrixStats’:\n",
      "\n",
      "    colMaxs, colMins, colRanges, rowMaxs, rowMins, rowRanges\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    aperm, apply, rowsum\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘Seurat’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:SummarizedExperiment’:\n",
      "\n",
      "    Assays\n",
      "\n",
      "\n",
      "Package 'mclust' version 5.4.6\n",
      "Type 'citation(\"mclust\")' for citing this R package in publications.\n",
      "\n",
      "\n",
      "Attaching package: ‘mclust’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:maps’:\n",
      "\n",
      "    map\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘Matrix’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:S4Vectors’:\n",
      "\n",
      "    expand\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:reshape’:\n",
      "\n",
      "    expand\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "library(SymSim)\n",
    "library(SingleCellExperiment)\n",
    "library(Seurat)\n",
    "\n",
    "library(mclust)\n",
    "library(scran)\n",
    "library(Matrix)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAA0gAAANICAMAAADKOT/pAAACoFBMVEUAAAABAQECAgIDAwME\nBAQFBQUGBgYHBwcICAgJCQkKCgoLCwsMDAwNDQ0ODg4PDw8RERESEhITExMUFBQVFRUYGBgZ\nGRkaGhobGxscHBwdHR0eHh4fHx8gICAhISEiIiIjIyMkJCQlJSUmJiYnJycoKCgqKiorKyss\nLCwtLS0uLi4vLy8wMDAxMTEyMjIzMzM0NDQ1NTU2NjY3Nzc5OTk6Ojo7Ozs8PDw9PT0+Pj4/\nPz9AQEBBQUFCQkJDQ0NERERFRUVGRkZHR0dISEhJSUlKSkpMTExNTU1PT09QUFBSUlJTU1NU\nVFRVVVVWVlZXV1dZWVlaWlpbW1tcXFxdXV1eXl5fX19gYGBhYWFiYmJjY2NkZGRlZWVmZmZn\nZ2doaGhpaWlqampra2tsbGxtbW1ubm5vb29wcHBxcXFzc3N0dHR1dXV2dnZ3d3d4eHh5eXl6\nenp8fHx9fX1+fn6BgYGCgoKDg4OFhYWHh4eIiIiJiYmKioqLi4uNjY2Ojo6Pj4+QkJCRkZGS\nkpKTk5OUlJSVlZWWlpaXl5eYmJiZmZmampqcnJydnZ2enp6fn5+goKChoaGioqKjo6OkpKSl\npaWmpqanp6eoqKipqamurq6vr6+wsLCxsbGzs7O1tbW2tra3t7e4uLi5ubm6urq7u7u8vLy9\nvb2+vr6/v7/AwMDBwcHCwsLExMTFxcXGxsbIyMjJycnKysrLy8vNzc3Ozs7Pz8/Q0NDR0dHS\n0tLU1NTV1dXW1tbX19fY2NjZ2dna2trc3Nze3t7f39/g4ODh4eHj4+Pk5OTl5eXm5ubn5+fp\n6enq6urr6+vs7Ozu7u7v7+/w8PDx8fHy8vLz8/P09PT19fX29vb39/f4+Pj5+fn6+vr7+/v8\n/Pz9/f3+/v7///86JRXaAAAACXBIWXMAABJ0AAASdAHeZh94AAATtUlEQVR4nO3d+7flZUHH\n8WcuIMooA4qiSWDekCQ0L0OOOhaZWsrNC2IwZSUIaY1WFpmpaRczEkWDMUtN0kxLM0lJEczQ\nLFORvDsz/0rMnKHWggfWejifZ5/v9/u8Xr/s56fvZ3/n7Pda++x91ppyAFi3stFPAJZASBAg\nJAgQEgQICQKEBAFCggAhQYCQIEBIECAkCBASBAgJAoQEAUKCACFBgJAgQEgQICQIEBIECAkC\nhAQBQoIAIUGAkCBASBAgJAgQEgQICQKEBAFCggAhQYCQIEBIECAkCBASBAgJAoQEAUKCACFB\ngJAgQEgQICQIEBIECAkChAQBQoIAIUGAkCBASBAgJAgQEgQICQKEBAFCggAhQYCQIEBIECAk\nCBASBAgJAoQEAUKCACFBgJAgQEgQICQIEBIECAkChAQBQoIAIUGAkCBASBAgJAgQEgQICQKE\nBAFCggAhQYCQIEBIECAkCBASBAgJAoQEAUKCACFBgJAgQEgQICQIEBIECAkChAQBQoIAIUGA\nkCBASBAgJAgQEgQICQKEBAFCggAhQYCQIEBIECAkCBASBAgJAoQEAUKCACFBgJAgQEgs360n\nlGd3nhASy3dp2fSQzhNCYvH+9Ygzf7R8pe+GkFi8p279zAvKB/puCImle2d52YHXlsv7jgiJ\nhfvGg4776oH3l7P6rmxkSNddcy10dfOBAy8vbz5w4EvlR/q+mDcypNPvtR262nPg+q2P/sFt\nL7bjNt3S9cW8kSGdf/4GjjOIp6x9zLCz/H3XGSGxaO8ot3t91x0hsWS3HH/EBS856OnlhV2H\nhMSSXVxesXb4RDml65CQWLBPb33I/6ydvr1ly7d6LgmJBTuj7L39+MjysZ5LQoIAIUGAkCBA\nSBAgJAgQEgQICQKEBAFCggAhQYCQIEBIECAkCBASBAgJAoQEAUKCACFBgJAgQEgQICQIEBIE\nCAkChAQBQoIAIUGAkCBASBAgJAgQEgQICQKEBAFCggAhQYCQIEBIECAkCBASBAgJAoQEAUKC\nACFBgJAgQEgQICQIEBIECAkChAQBQoIAIUGAkCBASBAgJAgQEgQICQKEBAFCggAhQYCQWLhd\n5Tabj3vGh7quCImFO3bL7t27z3tE2fqpnitCYtluKo89+LDvyeV3e84IiWW7uvziocdzy9t7\nzgiJZbtsLaB/uu8Jt/ScERLLtrOcc/HFFz5ly1Nu7DqzlJA+d/pFcCd79x9TDjnpnbkXW81S\nQnpP2egfGVO094ay47aXxy1/c0J5d+7VVrGUkD5c9ucuxnJcVX7l0OMflud23RESi3ZJueLQ\n4zVlV9cdIbFoO8p1hx4vKb/UdUdILNm+bUd+9+Dju4/o+4cNQmLRri/3271799mnlKPe0XdI\nSCzZlYc++z76lItv6jwkJAgQEgQICQKEBAFCggAhQYCQIEBIECAkCBASBAgJAoQEAUKCACFB\ngJAgQEgQICQIEBIECAkChAQBQoIAIUGAkCBASBAgJAgQEgQICQKEBAFCggAhQYCQIEBIECAk\nCBASBAgJAoQEAUKCACFBgJAgQEgQICQIEBIECAkChAQBQoIAIUGAkCBASBAgJAgQEgQICQKE\nBAFCggAhQYCQIEBIECAkCBASBAiJxbvx137s/lvv//Q//nbHDSGxcPtffWR55PN+/hn3Lmd0\nXBESC3d++eEPH3z82mXndlwREsv2unLyfx8+fr3jjJBYtP84ast1q9jZyJCe+ZhXxpwrJGpe\nUV60kp2NDOkXHr0r5vRkSPfftp1l2PPw8qHcC+NubGRISdG3dg++9FqW4Yay6Vu5F8bdEFLF\nSX+WuxYb6vPlfqsZElKFkBbjP8sR31/JkJAqhLQcJ5b3HT71DUpIFUJajjeUh3784ON33vkz\nXXeEVCGk5dh/Qdn0hOe/eNd9y3ldd4RUIaQl+evnPOiI7Y866+1f7boipAoh0UpIFUKilZAq\nhEQrIVUIiVZCqhASrYRUISRaCalCSLQSUoWQaCWkCiHRSkgVQqKVkCqERCshVQiJVkKqEBKt\nhFQhJFoJqUJItBJShZBoJaQKIdFKSBVCopWQKoREKyFVCIlWQqoQEq2EVCEkWgmpQki0ElKF\nkGglpAoh0UpIFUKilZAqhEQrIVUIiVZCqhASrYRUISRaCalCSLQSUoWQaCWkCiHRSkgVQqKV\nkCqERCshVQiJVkKqEBKthFQhJFoJqUJItBJShZBoJaQKIdFKSBVCopWQKoREKyFVCIlWQqoQ\nEq2EVCEkWgmpQki0ElKFkGglpAoh0UpIFUKilZAqhEQrIVUIiVZCqhASrYRUISRaCalCSLQS\nUoWQaCWkCiHRSkgVQlqSXaWUTdsedvZfBl8hdyakCiEtybHlot0XnXPapvK0b3RcEVKFkBbk\npnLyocfPnV6e13FGSBVCWpCry9lrhy/dq3y+34yQKoS0IJeV3zt82lne1W9mKSF9oPzJW2Ie\n8MaNvh1idpa/O3w6q/xRv5mlhPSRbSflHPHruSf2hSuvZePcvP+YTbd/xnBmuSb3c72jpYQU\nlXxrd/bW7WycPTeUR97+ozix/Evu53pHQqpIhvSaHblr0e6q8vzDp8+WB3XcEVKFkJbjkvKG\nw6fzym933BFShZCWY0f56NrhzeVEX8iumJAWY9+2Ld88+Pj588v2T/UcElKFkBbj+nL07t0X\nPvdRpTy547exB4RUJaTFuLLc5t4PfNLL/6HzkJAqhEQrIVUIiVZCqhASrYRUISRaCalCSLQS\nUoWQaCWkCiHRSkgVQqKVkCqERCshVQiJVkKqEBKthFQhJFoJqUJItBJShZBoJaQKIdFKSBVC\nopWQKoREKyFVCIlWQqoQEq2EVCEkWgmpQki0ElKFkGglpAoh0UpIFUKilZAqhEQrIVUIiVZC\nqhASrYRUISRaCalCSLQSUoWQaCWkCiHRSkgVQqKVkCqERCshVQiJVkKqEBKthFQhJFoJqUJI\ntBJShZBoJaQKIdFKSBVCopWQKoREKyFVCIlWQqoQEq2EVCEkWgmpQki0ElKFkGglpAoh0UpI\nFUKilZAqhEQrIVUIiVZCqhASrYRUISRaCalCSLQSUoWQaCWkCiHRSkgVQqKVkCqERCshVQiJ\nVkKqENKS7CqH/WbPFSFVCGlJjt38ojUf77kipAohLchN5dRVzAipQkgLcnW5aBUzQqq47wNO\nitl+wkbfzeAuK3++ihkhVbzx9W+JefZjN/puBrez/NR5h9zadUZInSXf2n3z8RdcRJO9+485\n/JndcbmfQ42QOkuG9IVy9ka/MOdm7w1lNb+kCqmzZEhfLDfmLjaIq8ovr2RHSJ0JaWNdUoIf\nwd4NIXUmpI21o/zzSnaE1JmQNtS+bVu/s5IhIXUmpA11fTn68B8I/VvfISF1JqQNdeXtf7G6\nue/XSELqTUhjEFJnQhqDkDoT0hiE1JmQxiCkzoQ0BiF1JqQxCKkzIY1BSJ0JaQxC6kxIYxBS\nZ0Iag5A6E9IYhNSZkMYgpM6ENAYhdSakMQipMyGNQUidCWkMQupMSGMQUmdCGoOQOhPSGITU\nmZDGIKTOhDQGIXUmpDEIqTMhjUFInQlpDELqTEhjEFJnQhqDkDoT0hiE1JmQxiCkzoQ0BiF1\nJqQxCKkzIY1BSJ0JaQxC6kxIYxBSZ0Iag5A6E9IYhNSZkMYgpM6ENAYhdSakMQipMyGNQUid\nCWkMQupMSGMQUmdCGoOQOhPSGITUmZDGIKTOhDQGIXUmpDEIqTMhjUFInQlpDELqTEhjEFJn\nQhqDkDoT0hiE1JmQxiCkzoQ0BiF1JqQxCKkzIY1BSJ0JaQxC6kxIYxBSZ0Iag5A6E9JG21XW\nvKDripA6u/TEy2P2lM9u9O3M0LGbX3TIX3VdEVJnv3H86TGnlk9u9O3Mz03lMauYEdKMeGt3\nD1xdLlrFjJBmREj3wGXlilXMCGlGhHQP7CzPPPgb0sWdZ4Q0I9GQHn2f7QPYs/+Ytc/snpb7\nl6sS0oxEQzrjnGsHcPMN5Yzcv9ndENKMREP6yT25a03YVeVlK9kR0owIqd0l5W0r2RHSjAip\n3Y7y6ZXsCGlGhNRs37ajvr+SISHNiJCaXV8ev5ohIc2IkJpdWXavZkhIMyKk6RLSjAhpuoQ0\nI0KaLiHNiJCmS0gzIqTpEtKMCGm6hDQjQpouIc2IkKZLSDMipOkS0owIabqENCNCmi4hzYiQ\npktIMyKk6RLSjAhpuoQ0I0KaLiHNiJCmS0gzIqTpEtKMCGm6hDQjQpouIc2IkKZLSDMipOkS\n0owIabqENCNCmi4hzYiQpktIMyKk6RLSjAhpuoQ0I0KaLiHNiJCmS0gzIqTpEtKMCGm6hDQj\nQpouIc2IkKZLSDMipOkS0owIabqENCNCmi4hzYiQpktIMyKk6RLSjAhpuoQ0I0KaLiHNiJCm\nS0gzIqTpEtKMCGm6hDQjQpouIc2IkKZLSDMipOkS0owIabqENCNCmi4hzYiQpktIMyKk6RLS\njAhpuoQ0I0KaLiHNiJCmS0gzIqTpEtKMCGm6hDQjXyxvuzbmcZdu9O0sipBm5MubStCZG307\niyKkUXlrFyWkUQ0U0r+/6gnbt25/4qtu7LghpFGNE9LrjioP/rkLzzqpbLmm34iQRjVMSL9a\nfug9+w8ePnbGB/utCGlUo4R0RTn5i4eP+77Xb0ZIoxokpK9s3/KJVewIaVSDhPS68rMr2RHS\nqJIh3fLW3BfFWTc/ruw9+Ayv232bV+fu+E6ENKpkSK/ZvH2i9tynfPngM/ydg19BPyt3x3ci\npFElQ3rXA3PXyvpaKfsOH99UfqvjkJBGNUZI3ynl64ePLyjv7TgkpFGNEdKBU8tfHD49ovxX\nxx0hjWqQkN5ajj/08fe+D256aM8dIY1qkJAOvLRsesKLLzjzgWVrzw/thDSsUUI68N7nHL/1\n6JOf86YvdV0R0qiGCWk1hDQqIUUJaVRCihLSqIQUJaRRCSlKSKMSUpSQRiWkKCGNSkhRQhqV\nkKKENCohRQlpVEKKEtKohBQlpFEJKUpIoxJSlJBGJaQoIY1KSFFCGpWQooQ0KiFFCWlUQooS\n0qiEFCWkUQkpSkijElKUkEYlpCghjUpIUUIalZCihDQqIUUJaVRCihLSqIQUJaRRCSlKSKMS\nUpSQRiWkKCGNSkhRQhqVkKKENCohRQlpVEKKEtKohBQlpFEJKUpIoxJSlJBGJaQoIY1KSFFC\nGpWQooQ0KiFFCWlUQooS0qiEFCWkUQkpSkijElKUkEZ12hNfGfOsYzb6bjackEZ11mm7Yk45\neqPvZsMJifXz1k5IBAhJSAQISUgECElIBAhJSAQISUgECElIBAhJSAQISUgECElIBEw6pF3l\nw2uH3eVP+60IifWbdEjHbrl17fDY8sl+K0Ji/aYc0k3l1LXDN7fe63v9ZoTE+k05pKvLhWuH\nj5Qf7zgjJNZvyiFdVt66dnhteWnHGSGxflMOaWd5xnmHPKxc0XFGSKxfNKSnPv+inL37jyn/\n59PBZ3lHQmL9oiFt/elkSDeUJ65d9qvlPj8IPss7EhLrFw3pqPcFL3bgqtt/M/rb8qTkde9I\nSKzfhEO65P8/a3hZ8rp3JCTWb8Ih7bj9W9hzytuT170jIbF+0w1p37Yjv7t2ekT5TPC6dyIk\n1m+6IV1fTls73Lp5277gde9ESKzfdEO6srxk7fDR8hPBy96ZkFi/6Ya0MkJi/YQkJAKEJCQC\nhCQkAoQkJAKEJCQChCQkAoQkJAKEJCQChCQkAoQkJAKEJCQChCQkAoQkJAKEJCQChCQkAoQk\nJAKEJCQChCQkAoQkJAKEJCQChCQkAoQkJAKEJCQChCQkAoQkJAKEJCQChCQkAi4/8vSczT3/\nz+RuhMT63Xj+5Tlb92707dwTQmJivLWDACFBgJAgQEgQICQIEBIECAkChAQBQoIAIUGAkCBA\nSBAgJAgQEgQICQKEBAFCggAhQYCQIEBIECAkCBASBAgJAoQEAUKCACFBgJAgQEgQkA5pV/lQ\n9oJVQmJi0iEdu+nr2QtWCYmJCYd0U3l49Hp3QUhMTDikq8u50evdBSExMeGQLiu/H73eXRAS\nExMOaWf5SPR6d0FITMzmbdtz9uw/ZvOtq3jWQmJirn7/tTk331AetZJnLSQW7arywpXsCIlF\nu6T8wUp2hMSi7Sj/uJIdIbFk+7Zt/dZKhoTEkl1fTl3NkJBYsivLBasZEhIECAkChAQBQoIA\nIUGAkCBASBAgJAgQEgQICQKEBAFCggAhQYCQIEBIECAkCBASBAgJAoQEAUKCACFBgJAgQEgQ\nICQIEBIECAkChAQBQoIAIUGAkCBASBAgJAgQEgQICQKEBAFCggAhQYCQIEBIECAkCBASBAgJ\nAoQEAUKCACFBgJAgQEgQICQIEBIECAkChAQBQoIAIUGAkCBASBAgJAgQEgQICQKEBAFCggAh\nQYCQIEBIECAkCBASBAgJAoQEAUKCACFBgJAgQEgQICQIEBIECAkChAQBQoIAIUGAkCBASBAg\nJAgQEgQICQKEBAFCggAhQYCQIEBIECAkCBASBAgJAoQEAUKCACFBgJAgQEgQICQIEBIECAkC\nhAQBQoIAIUGAkCBASBAgJAgQEgQICQKEBAFCggAhQYCQIEBIECAkCBASBAgJAoQEAUKCACFB\ngJAgQEgQICQIEBIECAkChAQBQoIAIUGAkCBASBAgJAgQEgQICQKEBAFCggAhQYCQIEBIECAk\nCBASBAgJAv4XryXLfAN9MckAAAAASUVORK5CYII=",
      "text/plain": [
       "plot without title"
      ]
     },
     "metadata": {
      "image/png": {
       "height": 420,
       "width": 420
      },
      "text/plain": {
       "height": 420,
       "width": 420
      }
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "phyla_disc <- read.tree('tree_sameend.txt')\n",
    "#par(mfrow=c(1,2))\n",
    "plot(phyla_disc)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "filter_exp <-function(sce,pct){\n",
    "    expsn <- rowMeans(logcounts(sce))\n",
    "    keep <- order(expsn, decreasing = TRUE)[seq_len(pct/100*length(expsn))]\n",
    "    sce_filtered <- sce[keep, ]  \n",
    "    \n",
    "    sce_filtered\n",
    "}\n",
    "\n",
    "save_sce <- function(sce,path,lowdim='TSNE'){\n",
    "    \n",
    "    sizeFactors(sce) <- calculateSumFactors(sce)\n",
    "    \n",
    "    save_data <- Matrix(as.matrix(assay(sce,'counts')),sparse=TRUE)\n",
    "    \n",
    "    writeMM(save_data,sprintf(\"%s.data.counts.mm\",path))\n",
    "    write.table(as.matrix(rownames(save_data)),sprintf('%s.data.row',path),row.names=FALSE, col.names=FALSE)\n",
    "    write.table(as.matrix(colnames(save_data)),sprintf('%s.data.col',path),row.names=FALSE, col.names=FALSE)\n",
    "    \n",
    "    tsne_data <- reducedDim(sce, lowdim)\n",
    "    colnames(tsne_data) <- c(sprintf('%s_1',lowdim),sprintf('%s_2',lowdim))\n",
    "    print(head(cbind(as.matrix(colData(sce)),tsne_data)))\n",
    "    write.table(cbind(as.matrix(colData(sce)),tsne_data),sprintf('%s.metadatacol.tsv',path),row.names=TRUE, col.names=TRUE,sep='\\t')    \n",
    "    write.table(cbind(as.matrix(rowData(sce))),sprintf('%s.metadatarow.tsv',path),row.names=TRUE, col.names=TRUE,sep='\\t')    \n",
    "    \n",
    "    write.table(sizeFactors(sce),file=sprintf('%s.size_factor.tsv',path),sep='\\t',row.names=FALSE, col.names=FALSE)    \n",
    "    \n",
    "}\n",
    "\n",
    "save_hvg_seurat <-function(sce,path,method){\n",
    "    seurat <- CreateSeuratObject(counts = counts(sce), project = \"scRNAseq\", assay = \"RNA\",\n",
    "                                         min.cells = 0, min.features = 0,\n",
    "                                         meta.data = as.data.frame(colData(sce)))\n",
    "    #seurat <- ScaleData(seurat, vars.to.regress = \"nUMI\", display.progress = FALSE)\n",
    "    seurat <- NormalizeData(seurat)\n",
    "    seurat <- FindVariableFeatures(seurat, selection.method = method, nfeatures = length(rownames(seurat@assays$RNA)))\n",
    "    vf <- VariableFeatures(seurat)  \n",
    "    write.table(vf,file=sprintf('%s.hvg.%s.tsv',path,method),sep='\\t',row.names=FALSE, col.names=TRUE)       \n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Centering and scaling data matrix\n",
      "\n",
      "PC_ 1 \n",
      "Positive:  1438, 374, 2052, 4221, 2215, 3800, 4882, 4898, 3534, 4883 \n",
      "\t   4684, 3092, 2563, 3481, 2803, 2885, 2881, 3019, 4241, 2574 \n",
      "\t   3454, 3565, 3719, 2867, 4330, 1158, 1736, 1562, 1929, 2135 \n",
      "Negative:  2905, 2646, 4396, 2477, 4516, 3824, 4645, 4851, 2971, 4832 \n",
      "\t   3452, 4231, 2601, 82, 1866, 604, 454, 1501, 2414, 1259 \n",
      "\t   1814, 1480, 1521, 1083, 641, 3311, 3303, 630, 1298, 1696 \n",
      "PC_ 2 \n",
      "Positive:  2240, 4776, 1516, 3269, 2696, 4953, 3961, 4321, 4083, 809 \n",
      "\t   1344, 4841, 4899, 3519, 3759, 1259, 1814, 2232, 2857, 3890 \n",
      "\t   3638, 4468, 4080, 2642, 1298, 3977, 3464, 4729, 4193, 4096 \n",
      "Negative:  2052, 4684, 374, 1438, 3534, 2215, 4882, 4883, 2563, 3019 \n",
      "\t   2881, 4221, 2885, 4898, 3800, 3481, 3454, 2803, 3092, 2337 \n",
      "\t   1736, 2135, 1158, 4241, 1562, 3565, 1929, 2484, 3719, 2867 \n",
      "PC_ 3 \n",
      "Positive:  4575, 3303, 358, 1696, 3311, 4271, 4805, 2624, 3072, 4658 \n",
      "\t   628, 630, 3806, 256, 1760, 3797, 930, 3893, 1618, 4985 \n",
      "\t   456, 1980, 1504, 825, 2550, 4988, 45, 433, 329, 561 \n",
      "Negative:  1516, 2240, 2696, 4776, 3961, 4096, 4083, 3269, 4321, 606 \n",
      "\t   4953, 1675, 3519, 1936, 1149, 856, 2097, 712, 2299, 1824 \n",
      "\t   1969, 1315, 2193, 4836, 983, 3534, 1941, 1134, 4080, 3454 \n",
      "PC_ 4 \n",
      "Positive:  2240, 4776, 2696, 3961, 4321, 3269, 4953, 3519, 4575, 3303 \n",
      "\t   1299, 3198, 1315, 358, 2624, 3311, 1420, 1696, 4805, 2499 \n",
      "\t   2097, 435, 456, 1798, 4080, 1173, 171, 630, 110, 1969 \n",
      "Negative:  1344, 4841, 809, 4899, 4468, 2857, 3638, 2642, 3013, 3977 \n",
      "\t   3890, 52, 3759, 593, 2056, 360, 3148, 909, 1327, 3464 \n",
      "\t   1091, 169, 3984, 4193, 2469, 675, 4836, 64, 3797, 2355 \n",
      "PC_ 5 \n",
      "Positive:  4836, 3961, 1696, 4776, 3269, 2240, 2696, 3820, 3913, 358 \n",
      "\t   3303, 750, 1824, 3311, 116, 1235, 2193, 607, 4321, 2726 \n",
      "\t   4658, 2710, 630, 967, 583, 2290, 4953, 1672, 295, 459 \n",
      "Negative:  2646, 2477, 4851, 3452, 4832, 4645, 2905, 3824, 4396, 4516 \n",
      "\t   2971, 4231, 1551, 2601, 1149, 4174, 82, 2414, 1491, 894 \n",
      "\t   1251, 2075, 3627, 1866, 1083, 1076, 1521, 2476, 2255, 2427 \n",
      "\n",
      "Computing nearest neighbor graph\n",
      "\n",
      "Computing SNN\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Modularity Optimizer version 1.3.0 by Ludo Waltman and Nees Jan van Eck\n",
      "\n",
      "Number of nodes: 1000\n",
      "Number of edges: 68622\n",
      "\n",
      "Running Louvain algorithm...\n",
      "Maximum modularity in 10 random starts: 0.4150\n",
      "Number of communities: 10\n",
      "Elapsed time: 0 seconds\n",
      "[1] 0.5025691\n",
      "       orig.ident nCount_RNA nFeature_RNA phenoid RNA_snn_res.2 seurat_clusters\n",
      "cell_1          1      35289         2940       1             1               1\n",
      "cell_2          1      23946         2644       1             1               1\n",
      "cell_3          1      17523         2246       1             1               1\n",
      "cell_4          1      24996         2737       1             1               1\n",
      "cell_5          1      25488         2809       1             1               1\n",
      "cell_6          1      16807         2371       1             1               1\n",
      "       ident    TSNE_1     TSNE_2\n",
      "cell_1     1 -2.759957  2.0702528\n",
      "cell_2     1 -1.149316  1.6876215\n",
      "cell_3     1 -9.187028 -7.6431681\n",
      "cell_4     1 -6.757827  5.5681751\n",
      "cell_5     1 -6.363901  1.9040204\n",
      "cell_6     1 -4.696057  0.5662241\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Centering and scaling data matrix\n",
      "\n",
      "PC_ 1 \n",
      "Positive:  3738, 2764, 4461, 832, 1025, 2945, 3088, 2333, 4405, 96 \n",
      "\t   4401, 3056, 3617, 202, 3885, 4446, 2485, 1220, 164, 1521 \n",
      "\t   3635, 1268, 4731, 743, 4302, 2757, 4696, 4171, 1925, 2695 \n",
      "Negative:  2792, 972, 4931, 4929, 4954, 3521, 4284, 3860, 3893, 3693 \n",
      "\t   4570, 3153, 4430, 3956, 2782, 3768, 4381, 2732, 3726, 4385 \n",
      "\t   1719, 4856, 2557, 4264, 4291, 4293, 3516, 4549, 3928, 2678 \n",
      "PC_ 2 \n",
      "Positive:  2973, 2946, 1905, 3621, 3174, 1025, 832, 3617, 3736, 4605 \n",
      "\t   3388, 96, 4654, 4737, 2945, 3008, 2601, 2597, 3088, 4405 \n",
      "\t   4446, 2724, 2702, 1060, 4316, 2779, 4302, 3635, 4326, 2578 \n",
      "Negative:  2757, 1268, 4731, 164, 2333, 2944, 3144, 3187, 1840, 1912 \n",
      "\t   2625, 2924, 2695, 222, 470, 2594, 4171, 4856, 4570, 347 \n",
      "\t   4498, 4264, 3768, 3302, 4291, 3486, 4385, 4381, 3726, 304 \n",
      "PC_ 3 \n",
      "Positive:  1268, 164, 4731, 3144, 4461, 4614, 470, 4415, 3187, 1840 \n",
      "\t   1127, 2508, 2625, 1849, 2944, 3501, 1912, 3738, 3049, 4343 \n",
      "\t   2764, 3240, 4572, 4872, 4676, 3072, 4693, 3848, 2362, 3662 \n",
      "Negative:  2333, 2724, 1060, 3008, 2968, 3433, 2578, 4751, 4326, 2719 \n",
      "\t   4316, 2702, 3621, 972, 4648, 2946, 2973, 2792, 4929, 3174 \n",
      "\t   4662, 4284, 1905, 2601, 3521, 4954, 3893, 4931, 3860, 3693 \n",
      "PC_ 4 \n",
      "Positive:  2973, 2946, 1905, 2601, 3174, 3140, 4605, 4654, 3388, 4737 \n",
      "\t   3621, 2597, 4570, 3736, 799, 153, 2414, 1325, 3768, 3726 \n",
      "\t   4498, 4856, 4291, 4385, 4381, 2340, 4264, 2333, 1483, 2148 \n",
      "Negative:  1060, 4326, 2719, 2968, 2724, 4316, 4648, 4751, 2578, 4662 \n",
      "\t   3008, 2702, 3877, 3433, 1268, 164, 4731, 1689, 1838, 1469 \n",
      "\t   3854, 4060, 119, 4094, 1509, 3424, 996, 4575, 2448, 1840 \n",
      "PC_ 5 \n",
      "Positive:  3617, 1025, 832, 96, 2945, 3088, 4302, 4401, 3635, 4405 \n",
      "\t   4446, 4696, 2779, 202, 4731, 3885, 1268, 163, 164, 743 \n",
      "\t   2485, 2482, 54, 1220, 377, 2208, 717, 1602, 1935, 1561 \n",
      "Negative:  2944, 3144, 3187, 1912, 2946, 2973, 1905, 3174, 4605, 4737 \n",
      "\t   222, 3388, 4654, 2473, 4662, 2601, 4751, 2597, 2414, 754 \n",
      "\t   3877, 4648, 2968, 2719, 1483, 2151, 4326, 799, 4461, 3738 \n",
      "\n",
      "Computing nearest neighbor graph\n",
      "\n",
      "Computing SNN\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Modularity Optimizer version 1.3.0 by Ludo Waltman and Nees Jan van Eck\n",
      "\n",
      "Number of nodes: 1000\n",
      "Number of edges: 70101\n",
      "\n",
      "Running Louvain algorithm...\n",
      "Maximum modularity in 10 random starts: 0.4418\n",
      "Number of communities: 8\n",
      "Elapsed time: 0 seconds\n",
      "[1] 0.5467303\n",
      "       orig.ident nCount_RNA nFeature_RNA phenoid RNA_snn_res.2 seurat_clusters\n",
      "cell_1          1      18371         2376       1             1               1\n",
      "cell_2          1      23824         2187       1             3               3\n",
      "cell_3          1      30422         2738       1             3               3\n",
      "cell_4          1      40380         2793       1             3               3\n",
      "cell_5          1      19174         2492       1             1               1\n",
      "cell_6          1      18315         2448       1             3               3\n",
      "       ident     TSNE_1     TSNE_2\n",
      "cell_1     1 -0.8652076  4.7878377\n",
      "cell_2     3  5.6390738  3.2916493\n",
      "cell_3     3 11.7686630 -0.6114393\n",
      "cell_4     3  4.5127196  0.6179032\n",
      "cell_5     1 -4.4665360  2.2281098\n",
      "cell_6     3  7.0213730 -0.5112917\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Centering and scaling data matrix\n",
      "\n",
      "PC_ 1 \n",
      "Positive:  1885, 2594, 4003, 1980, 4185, 1268, 4018, 2676, 2786, 3797 \n",
      "\t   2675, 3533, 3287, 4273, 2991, 3475, 3545, 2720, 397, 1157 \n",
      "\t   1609, 1994, 407, 1965, 1162, 1320, 61, 532, 2075, 2318 \n",
      "Negative:  4700, 1259, 1190, 3532, 2253, 3389, 4709, 3298, 3546, 4943 \n",
      "\t   4479, 2820, 66, 2557, 3290, 1648, 1304, 4885, 2317, 3908 \n",
      "\t   139, 49, 2048, 3229, 1947, 4793, 2913, 1668, 4518, 1935 \n",
      "PC_ 2 \n",
      "Positive:  1213, 327, 3606, 857, 2730, 4619, 4433, 2814, 3510, 3450 \n",
      "\t   4994, 3966, 2671, 4078, 2759, 4499, 4346, 4302, 4617, 4100 \n",
      "\t   3457, 467, 4551, 2265, 1752, 109, 1679, 1234, 4024, 1977 \n",
      "Negative:  393, 2801, 3353, 4161, 69, 2833, 4163, 2277, 4514, 4809 \n",
      "\t   4782, 4052, 3703, 1885, 4263, 4018, 1980, 4185, 4324, 978 \n",
      "\t   3533, 2594, 3062, 4003, 2675, 2676, 3287, 3545, 4651, 1441 \n",
      "PC_ 3 \n",
      "Positive:  2277, 3703, 3062, 4499, 2730, 1415, 3966, 4302, 3450, 4994 \n",
      "\t   1213, 1538, 4433, 4078, 1752, 4024, 109, 1611, 4346, 1679 \n",
      "\t   3784, 3903, 1212, 1956, 2265, 3907, 4681, 574, 467, 1005 \n",
      "Negative:  3606, 327, 857, 4619, 3510, 2814, 4100, 2759, 2671, 3457 \n",
      "\t   2801, 2968, 393, 1007, 4161, 4551, 69, 4052, 4514, 1525 \n",
      "\t   2833, 4782, 4163, 547, 660, 2095, 4324, 1621, 4263, 1848 \n",
      "PC_ 4 \n",
      "Positive:  327, 857, 3606, 3510, 2759, 2671, 4100, 2814, 3457, 4619 \n",
      "\t   4551, 393, 2801, 4161, 4052, 3532, 69, 660, 2833, 1621 \n",
      "\t   3298, 4700, 4163, 4518, 1525, 2095, 4324, 1190, 1007, 1259 \n",
      "Negative:  2277, 3703, 2968, 3062, 4745, 1531, 1415, 3714, 1538, 3161 \n",
      "\t   911, 3353, 83, 4895, 3784, 2111, 4809, 196, 2046, 2009 \n",
      "\t   1104, 2722, 241, 2000, 927, 3047, 1619, 1488, 188, 1750 \n",
      "PC_ 5 \n",
      "Positive:  4700, 1213, 1259, 3389, 1190, 4709, 3532, 2253, 1885, 3298 \n",
      "\t   1980, 2594, 3966, 4003, 4433, 3450, 3287, 2676, 4018, 2557 \n",
      "\t   4185, 3546, 2730, 4994, 2720, 4617, 4346, 3533, 4078, 3475 \n",
      "Negative:  3353, 4809, 756, 360, 978, 69, 4782, 2111, 2671, 4651 \n",
      "\t   393, 343, 4263, 4886, 951, 4008, 3457, 327, 4100, 2112 \n",
      "\t   2722, 4985, 2759, 2814, 2801, 32, 4052, 857, 4163, 2833 \n",
      "\n",
      "Computing nearest neighbor graph\n",
      "\n",
      "Computing SNN\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Modularity Optimizer version 1.3.0 by Ludo Waltman and Nees Jan van Eck\n",
      "\n",
      "Number of nodes: 1000\n",
      "Number of edges: 77542\n",
      "\n",
      "Running Louvain algorithm...\n",
      "Maximum modularity in 10 random starts: 0.3527\n",
      "Number of communities: 13\n",
      "Elapsed time: 0 seconds\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "1 singletons identified. 12 final clusters.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] 0.5465665\n",
      "       orig.ident nCount_RNA nFeature_RNA phenoid RNA_snn_res.2 seurat_clusters\n",
      "cell_1          1      14869         2286       1             1               1\n",
      "cell_2          1      44446         2815       1             1               1\n",
      "cell_3          1      20773         2619       1             1               1\n",
      "cell_4          1      21838         2529       1             1               1\n",
      "cell_5          1      18130         2356       1             1               1\n",
      "cell_6          1      34986         3090       1             1               1\n",
      "       ident    TSNE_1    TSNE_2\n",
      "cell_1     1  9.803143 -7.813966\n",
      "cell_2     1  4.868114 -7.017155\n",
      "cell_3     1  5.548431 -5.279466\n",
      "cell_4     1 -4.226938  3.320177\n",
      "cell_5     1  7.876984 -7.612988\n",
      "cell_6     1  2.587866 -6.053874\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Centering and scaling data matrix\n",
      "\n",
      "PC_ 1 \n",
      "Positive:  3750, 3317, 1742, 3930, 4045, 906, 2923, 4923, 3268, 2612 \n",
      "\t   3213, 4080, 4467, 3167, 4082, 4401, 2698, 4740, 4386, 3218 \n",
      "\t   3940, 3601, 2266, 1494, 1389, 927, 1024, 233, 198, 2483 \n",
      "Negative:  2224, 1485, 3337, 3254, 4887, 2458, 3426, 3639, 2296, 2587 \n",
      "\t   4613, 4808, 3677, 2943, 4878, 3376, 2223, 4001, 4516, 4625 \n",
      "\t   4164, 4852, 4461, 3324, 3432, 3076, 2607, 2072, 2902, 3835 \n",
      "PC_ 2 \n",
      "Positive:  298, 2624, 4616, 3341, 1064, 4162, 2990, 2722, 4237, 4719 \n",
      "\t   2946, 3895, 2763, 4809, 4993, 3254, 2224, 1485, 3337, 4146 \n",
      "\t   3672, 3376, 3426, 4516, 2587, 4309, 4613, 4932, 4001, 3569 \n",
      "Negative:  3067, 1340, 4723, 3162, 2883, 3447, 2223, 4808, 2943, 2296 \n",
      "\t   2607, 4403, 1742, 3837, 2458, 4386, 4887, 3835, 2479, 3940 \n",
      "\t   2612, 588, 3639, 4026, 4878, 2241, 4522, 1712, 2396, 917 \n",
      "PC_ 3 \n",
      "Positive:  947, 1535, 2072, 1301, 3930, 4401, 2698, 3268, 1742, 1340 \n",
      "\t   4386, 1494, 1677, 692, 3455, 3761, 3162, 3067, 4790, 2244 \n",
      "\t   3447, 3554, 2948, 2050, 1493, 453, 3940, 2612, 1654, 958 \n",
      "Negative:  4045, 3750, 3317, 906, 2923, 3213, 4923, 4464, 4082, 3218 \n",
      "\t   3601, 4467, 3167, 4740, 2266, 4080, 927, 198, 1024, 331 \n",
      "\t   1689, 2483, 1389, 454, 2458, 1733, 4887, 2223, 2296, 4808 \n",
      "PC_ 4 \n",
      "Positive:  947, 1535, 2072, 4529, 4559, 4087, 2763, 4616, 422, 1591 \n",
      "\t   1869, 867, 1531, 748, 2058, 1301, 1064, 136, 1586, 19 \n",
      "\t   923, 2604, 1220, 3067, 528, 2268, 2455, 3322, 59, 1038 \n",
      "Negative:  4464, 4740, 4923, 4045, 3984, 906, 3167, 4080, 4082, 3218 \n",
      "\t   3213, 3324, 3317, 2948, 454, 3750, 2483, 3677, 1689, 3455 \n",
      "\t   2923, 3376, 927, 1677, 2833, 1733, 4625, 1190, 1485, 2164 \n",
      "PC_ 5 \n",
      "Positive:  4616, 2722, 3895, 3639, 1064, 2990, 3672, 4719, 298, 4162 \n",
      "\t   4887, 2458, 2946, 2624, 4237, 4809, 2296, 2763, 2943, 1798 \n",
      "\t   4808, 3745, 2419, 2314, 1677, 1561, 330, 851, 3051, 4026 \n",
      "Negative:  4529, 4559, 3930, 4635, 2604, 2698, 3268, 3322, 4401, 2587 \n",
      "\t   1494, 4087, 4164, 1409, 2484, 3223, 4613, 3677, 4472, 3376 \n",
      "\t   3205, 3426, 3324, 4516, 2959, 832, 2222, 2210, 1485, 4461 \n",
      "\n",
      "Computing nearest neighbor graph\n",
      "\n",
      "Computing SNN\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Modularity Optimizer version 1.3.0 by Ludo Waltman and Nees Jan van Eck\n",
      "\n",
      "Number of nodes: 1000\n",
      "Number of edges: 75537\n",
      "\n",
      "Running Louvain algorithm...\n",
      "Maximum modularity in 10 random starts: 0.4406\n",
      "Number of communities: 8\n",
      "Elapsed time: 0 seconds\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "1 singletons identified. 7 final clusters.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] 0.4640544\n",
      "       orig.ident nCount_RNA nFeature_RNA phenoid RNA_snn_res.2 seurat_clusters\n",
      "cell_1          1      26763         2767       1             3               3\n",
      "cell_2          1       8180         1709       1             3               3\n",
      "cell_3          1      20482         2621       1             3               3\n",
      "cell_4          1      30164         2874       1             3               3\n",
      "cell_5          1      23035         2605       1             3               3\n",
      "cell_6          1      15838         2231       1             3               3\n",
      "       ident   TSNE_1    TSNE_2\n",
      "cell_1     3 10.67308 -3.473428\n",
      "cell_2     3 12.22027 -4.602434\n",
      "cell_3     3 18.51620 -3.565648\n",
      "cell_4     3 15.44483 -2.345779\n",
      "cell_5     3 13.41163 -2.481266\n",
      "cell_6     3 13.65331 -1.775150\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Centering and scaling data matrix\n",
      "\n",
      "PC_ 1 \n",
      "Positive:  3154, 1832, 253, 3338, 3534, 2625, 3233, 2913, 2656, 4598 \n",
      "\t   3580, 1477, 2312, 4015, 1722, 1824, 2142, 516, 2852, 4955 \n",
      "\t   4553, 2822, 4602, 3439, 2729, 1001, 4582, 3108, 1377, 3100 \n",
      "Negative:  3005, 4181, 823, 4105, 3200, 4233, 3707, 4638, 4307, 2579 \n",
      "\t   3879, 1847, 2309, 2718, 4936, 3702, 3953, 4507, 4509, 4824 \n",
      "\t   3381, 4386, 2747, 490, 2537, 244, 4872, 2515, 76, 3075 \n",
      "PC_ 2 \n",
      "Positive:  490, 244, 3075, 2685, 3970, 2747, 4872, 3381, 4598, 2625 \n",
      "\t   3233, 4180, 2913, 2656, 1949, 1614, 425, 3828, 4087, 1796 \n",
      "\t   2208, 3252, 1892, 577, 1477, 516, 378, 1426, 4638, 744 \n",
      "Negative:  1847, 3702, 2309, 3573, 4241, 4710, 3680, 2556, 4553, 4083 \n",
      "\t   2729, 2902, 2852, 1001, 4955, 1016, 3439, 2945, 2822, 1884 \n",
      "\t   4849, 3607, 4602, 4028, 2145, 1377, 2291, 3308, 1747, 2247 \n",
      "PC_ 3 \n",
      "Positive:  244, 2685, 3075, 490, 2747, 4872, 3534, 3970, 253, 1832 \n",
      "\t   4180, 3828, 3252, 4087, 1796, 378, 577, 1614, 425, 2208 \n",
      "\t   1949, 2309, 1847, 4241, 693, 1426, 3381, 3702, 2142, 744 \n",
      "Negative:  823, 4553, 2729, 1001, 3879, 4936, 4638, 4509, 4507, 1377 \n",
      "\t   1884, 4044, 3010, 760, 2515, 1276, 611, 2864, 952, 339 \n",
      "\t   1997, 4608, 1091, 761, 4598, 686, 2913, 2656, 3233, 2054 \n",
      "PC_ 4 \n",
      "Positive:  3025, 4314, 3015, 3462, 3014, 1832, 253, 2122, 3702, 4710 \n",
      "\t   4241, 2309, 4760, 3534, 4955, 213, 210, 3712, 2852, 1390 \n",
      "\t   1847, 3573, 4602, 1016, 267, 1105, 3975, 3884, 1963, 1293 \n",
      "Negative:  4509, 823, 4936, 4638, 2515, 3879, 4994, 4507, 4968, 4623 \n",
      "\t   760, 4203, 4608, 2912, 3138, 4044, 3255, 313, 1997, 3433 \n",
      "\t   2935, 3590, 2884, 4429, 2830, 3010, 2468, 339, 3489, 3381 \n",
      "PC_ 5 \n",
      "Positive:  823, 4955, 3439, 2822, 2852, 4602, 4936, 4638, 1268, 4509 \n",
      "\t   2864, 4507, 3879, 4849, 339, 3108, 445, 3100, 2133, 4582 \n",
      "\t   2054, 760, 952, 3010, 1985, 761, 4666, 3138, 1997, 4044 \n",
      "Negative:  3534, 253, 1832, 3200, 4105, 4181, 4233, 2718, 3005, 3953 \n",
      "\t   2579, 3707, 4307, 3202, 2142, 722, 1368, 1493, 277, 273 \n",
      "\t   3025, 2122, 3014, 2537, 3712, 4386, 4353, 4314, 76, 4968 \n",
      "\n",
      "Computing nearest neighbor graph\n",
      "\n",
      "Computing SNN\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Modularity Optimizer version 1.3.0 by Ludo Waltman and Nees Jan van Eck\n",
      "\n",
      "Number of nodes: 1000\n",
      "Number of edges: 88124\n",
      "\n",
      "Running Louvain algorithm...\n",
      "Maximum modularity in 10 random starts: 0.2202\n",
      "Number of communities: 19\n",
      "Elapsed time: 0 seconds\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2 singletons identified. 17 final clusters.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] 0.4096234\n",
      "       orig.ident nCount_RNA nFeature_RNA phenoid RNA_snn_res.2 seurat_clusters\n",
      "cell_1          1      29312         3011       1             2               2\n",
      "cell_2          1      18841         2577       1            12               5\n",
      "cell_3          1      17003         2351       1            11               4\n",
      "cell_4          1      20562         2674       1             3              11\n",
      "cell_5          1      17599         2535       1            14               7\n",
      "cell_6          1      23691         2842       1            12               5\n",
      "       ident    TSNE_1    TSNE_2\n",
      "cell_1     2 1.8170596 -9.025335\n",
      "cell_2     5 0.6874378 -6.917772\n",
      "cell_3     4 7.0956123 -5.115148\n",
      "cell_4    11 3.3640085 -3.897435\n",
      "cell_5     7 8.3785148 -5.698294\n",
      "cell_6     5 1.5425780 -7.659840\n"
     ]
    }
   ],
   "source": [
    "for(i in c(6:10)){\n",
    "    #############\n",
    "    #True counts\n",
    "    #############\n",
    "    disc_true_counts <- SimulateTrueCounts(ncells_total=1000, \n",
    "                                          min_popsize=50, \n",
    "                                          i_minpop=1, \n",
    "                                          ngenes=5000, \n",
    "                                          nevf=20, \n",
    "                                          evf_type=\"discrete\", \n",
    "                                          n_de_evf=5,\n",
    "                                          vary=\"s\",\n",
    "                                          geffect_mean=1,\n",
    "                                          Sigma=0.4,\n",
    "                                          bimod = 1,\n",
    "                                          phyla=phyla_disc,\n",
    "                                          gene_effect_prob=0.01,\n",
    "                                          randseed=i)\n",
    "    \n",
    "    if(FALSE){\n",
    "        disc_true_counts_tsne <- PlotTsne(meta=disc_true_counts[[3]][,c('cellid','pop')],\n",
    "                                     data=log2(disc_true_counts[[1]]+1),\n",
    "                                     evf_type=\"discrete\",\n",
    "                                     n_pc=30,\n",
    "                                     label='pop',\n",
    "                                     saving = F,\n",
    "                                     plotname=\"discrete populations (true counts)\")\n",
    "        disc_true_counts_tsne[[2]]        \n",
    "    }\n",
    "    \n",
    "    \n",
    "    #############\n",
    "    #Observed counts\n",
    "    #############    \n",
    "    \n",
    "    gene_len <- sample(gene_len_pool, 5000, replace = FALSE)\n",
    "    \n",
    "    \n",
    "    disc_observed_UMI <- True2ObservedCounts(true_counts=disc_true_counts[[1]],\n",
    "                                             meta_cell=disc_true_counts[[3]],\n",
    "                                             protocol=\"UMI\",\n",
    "                                             alpha_mean=0.05,\n",
    "                                             alpha_sd=0.02,\n",
    "                                             gene_len=gene_len,\n",
    "                                             depth_mean=5e5,\n",
    "                                             depth_sd=3e4)   \n",
    "    \n",
    "    if(FALSE){\n",
    "        disc_observed_UMI_tsne <- PlotTsne(meta=disc_observed_UMI[[2]],\n",
    "                                           data=log2(disc_observed_UMI[[1]]+1),\n",
    "                                           evf_type=\"discrete\",\n",
    "                                           n_pc=20,\n",
    "                                           label='pop',\n",
    "                                           saving = F,\n",
    "                                           plotname=\"observed counts UMI\")\n",
    "        disc_observed_UMI_tsne[[2]]           \n",
    "    }\n",
    "    \n",
    "    #############\n",
    "    #Saving\n",
    "    #############    \n",
    "    \n",
    "    \n",
    "    count=disc_observed_UMI[[1]]\n",
    "    colnames(count)=c(disc_observed_UMI[[2]]['cellid'])$cellid\n",
    "    rownames(count)=c(1:dim(count)[1])\n",
    " \n",
    "    seurat <- CreateSeuratObject(counts = count)\n",
    "    seurat@meta.data$phenoid=c(disc_observed_UMI[[2]]['pop'])$pop    \n",
    "    \n",
    "    seurat=FindVariableFeatures(seurat, selection.method = \"vst\", nfeatures = 2000)\n",
    "\n",
    "    seurat_hvg <- NormalizeData(seurat)\n",
    "\n",
    "    all.genes <- rownames(seurat_hvg)\n",
    "    seurat_hvg=ScaleData(seurat_hvg,features = all.genes)\n",
    "\n",
    "    seurat_hvg_pca = RunPCA(seurat_hvg)\n",
    "\n",
    "    PC_cluster=50\n",
    "    PC_2D=50\n",
    "\n",
    "    seurat_hvg_pca <- RunTSNE(seurat_hvg_pca, dims= 1:PC_2D)\n",
    "\n",
    "    seurat_hvg_pca_label <- FindNeighbors(seurat_hvg_pca, dims = 1:PC_cluster)\n",
    "\n",
    "    seurat_hvg_pca_label_cluster <- FindClusters(seurat_hvg_pca_label, resolution = 2)\n",
    "\n",
    "    DimPlot(seurat_hvg_pca_label_cluster,group.by='phenoid',cols = DiscretePalette(22))\n",
    "\n",
    "    DimPlot(seurat_hvg_pca_label_cluster,group.by='seurat_clusters',cols = DiscretePalette(22))\n",
    "\n",
    "    print(adjustedRandIndex(seurat@meta.data$phenoid, seurat_hvg_pca_label_cluster@meta.data$seurat_clusters))\n",
    "\n",
    "    \n",
    "    sce_full_Simul <- as.SingleCellExperiment(seurat_hvg_pca_label_cluster)\n",
    "    \n",
    "    sce_full_Simul_expfiltered=filter_exp(sce_full_Simul,30)\n",
    "    \n",
    "    \n",
    "    path=sprintf('datasets/extract/Simul_1_1e-2_%s_filtered',i)\n",
    "    \n",
    "    save_sce(sce_full_Simul_expfiltered, path)\n",
    "\n",
    "    for(method in c('vst','mvp','disp')){\n",
    "        save_hvg_seurat(sce_full_Simul_expfiltered, path, method)\n",
    "    }    \n",
    "    \n",
    "    \n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "#help(SimulateTrueCounts)\n",
    "#0.01, 0.005, 0.001, 0.0005 0.0001"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "'datasets/extract/Kohinbulk_filtered'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "#saveRDS(disc_observed_UMI, file = \"datasets/simul/disc_observed_UMI12.rds\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "#saveRDS(disc_true_counts, file = \"datasets/simul/disc_true_counts12.rds\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "'datasets/extract/Kohinbulk_filtered'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "'datasets/extract/Simul_1_1e-2_5_filtered'"
      ],
      "text/latex": [
       "'datasets/extract/Simul\\_1\\_1e-2\\_5\\_filtered'"
      ],
      "text/markdown": [
       "'datasets/extract/Simul_1_1e-2_5_filtered'"
      ],
      "text/plain": [
       "[1] \"datasets/extract/Simul_1_1e-2_5_filtered\""
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message:\n",
      "“'normalizeSCE' is deprecated.\n",
      "Use 'logNormCounts' instead.\n",
      "See help(\"Deprecated\")”\n",
      "Warning message:\n",
      "“'centreSizeFactors' is deprecated.\n",
      "See help(\"Deprecated\")”\n"
     ]
    }
   ],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "       orig.ident phenoid    tSNE_1     tSNE_2\n",
      "cell_1          1       1  1.290648 -0.5960434\n",
      "cell_2          1       1  2.295489 -9.3229477\n",
      "cell_3          1       1  2.285761 -0.4733488\n",
      "cell_4          1       1 -3.107209  2.2224019\n",
      "cell_5          1       1 -2.828975  1.0804327\n",
      "cell_6          1       1  1.526465 -0.4216981\n"
     ]
    }
   ],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [],
   "source": [
    "sce_full_simul=SingleCellExperiment(assays = list(counts = count))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [],
   "source": [
    "colData(sce_full_simul)$phenoid=c(disc_observed_UMI[[2]]['pop'])$pop"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "class: SingleCellExperiment \n",
       "dim: 500 1000 \n",
       "metadata(0):\n",
       "assays(1): counts\n",
       "rownames(500): 1 2 ... 499 500\n",
       "rowData names(0):\n",
       "colnames(1000): cell_1 cell_2 ... cell_999 cell_1000\n",
       "colData names(0):\n",
       "reducedDimNames(0):\n",
       "spikeNames(0):\n",
       "altExpNames(0):"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sce_full_HumanLiver <- SingleCellExperiment(assays = list(counts = HumanLiverSeurat[['RNA']]@counts))"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.0.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
